Added debian packaging and upstart script

Shaywei opened the pull request:
Tested locally (including upstart functionality).

Commit:
Added debian packaging and upstart script

nresare commented:
I think we should create a server user on package install and have the upstart script drop privs into that user before we start.

Commit:
Do what the cool kids do

Put "9" in the debian/compat file.

Commit:
Add .tags to .gitignore

Used by Atom to navigate source code.

Commit:
Drop privileges before starting

Commit:
Start drserv on install, stop drserv before uninstall

Commit:
Set work directory to drserv user home directory

walles commented:
With the current incarnation you can now install this, have the user created and the service started.

It will use /home/drserv as its working directory.

I have not yet verified that it's actually possible to upload things.

Commit:
drserv-client: Make --key-file argument mandatory

The code crashes if called without that argument.

Commit:
Client: Explain key file format if trouble reading it

I had to do quite some digging to understand what format the key file was
supposed to be in.

This change encodes my findings in a hopefully helpful error message.

Commit:
Change default indexing command to "true"

It used to be "/usr/bin/true", which exists on OS X but not Linuces.

Just "true" works on Linux, and I'll just assume it works equally well on OS X.

Commit:
Make --auth-user mandatory

Not specifying it makes the client fail with a Traceback.

walles commented:
Now tested as well, I managed to actually deploy a package using the drserv-client I get through this package.

It strikes me however that the client and the server should really be in separate packages, will fix...

walles commented:
Suggestion: Take this in to begin with and I'll make splitting this into two into another pull request.

@nresare, can you please review this?

Commit:
Enable "setup.py flake8"

And fix the two warnings.

Commit:
Run flake8 as part of the build process

Olivia5k commented on setup.py:
```diff
@@ -7,9 +7,12 @@
 )
 
+setup_requires = (
+    'flake8',
```
I don't think your setup should require `flake8`. It's a linter tool and not a runtime dependency for your code. If you want to test with it, it should be in `tests_require`.

walles replied:
Source for that change:
http://flake8.readthedocs.org/en/latest/setuptools.html

So I think it should actually be there...

Olivia5k replied:
Interesting. I'd still argue that it should be in `tests_require`, but I can back down given that!

Olivia5k commented:
:+1: apart from my comment!

xeago commented on debian/upstart:
```diff
@@ -0,0 +1,33 @@
+  [ -f $DEFAULTFILE ] && . $DEFAULTFILE
+
+  # don't start unless explicitly enabled in config
+  if [ -z "$ENABLED" ] ; then
```
The current default of $ENABLED is `yes`, from there it is easily assumed that you can have it not enabled with `no`, testing for -z doesn't align with that assumption.

Olivia5k replied:
The current best practice for these things are, as I understand them, to do `ENABLED=true` (or `false` and then have the `if` clause look like `[[ -n "$ENABLED" ]] && $ENABLED`. That would catch if it is not set, but it can still explicitly be `true` or `false`.

Also, using `[[ ... ]]` over `[ ... ]` is also a best practice since `[` will start a new process and `[[` is the same thing but internal to the shell.

udoprog replied:
Be careful using `[[`, upstart uses /bin/sh for shells which does not guarantee the existence of the new test expressions.

`[[ -n "$ENABLED" ]] && $ENABLED` works with `true` because of it being a command, however it would also work with `ENABLED="ls"` which might not be desirable. A command executing with non-zero will suffice for the expression to be true.
The canonical, and portable /bin/sh test would be `[ "x$ENABLED" = "xtrue" ]`.

`[` should be a built-in just the same as `[[` in all modern shells

Olivia5k replied:
Ah, good. Thanks for the clarifications!

xeago commented on debian/default:
```diff
@@ -0,0 +1 @@
+ENABLED=yes
```
Why not switch this to ""? In that case you don't the problem of the package being installed on clients running you a server.

Commit:
Disable server by default

It can be enabled by putting a "yes" in /etc/default/drserv.

Also, the format of /etc/default/drserv has now been documented within that file
itself.
